src/gt4py/next/program_processors/runners/dace_fieldview/gtir_to_sdfg.py

```diff
@@ -281,93 +313,140 @@ def visit_Program(self, node: gtir.Program) -> dace.SDFG:
         else:
             head_state = entry_state
 
-        self._add_sdfg_params(sdfg, node.params)
+        sdfg_arg_names = self._add_sdfg_params(sdfg, node.params)
 
         # visit one statement at a time and expand the SDFG from the current head state
         for i, stmt in enumerate(node.body):
             # include `debuginfo` only for `ir.Program` and `ir.Stmt` nodes: finer granularity would be too messy
             head_state = sdfg.add_state_after(head_state, f"stmt_{i}")
             head_state._debuginfo = dace_common_util.debug_info(stmt, default=sdfg.debuginfo)
-            self.visit(stmt, sdfg=sdfg, state=head_state)
+            head_state = self.visit(stmt, sdfg=sdfg, state=head_state)
 
         # Create the call signature for the SDFG.
         #  Only the arguments required by the GT4Py program, i.e. `node.params`, are added
         #  as positional arguments. The implicit arguments, such as the offset providers or
         #  the arguments created by the translation process, must be passed as keyword arguments.
-        sdfg.arg_names = [str(a) for a in node.params]
+        sdfg.arg_names = sdfg_arg_names
 
         sdfg.validate()
         return sdfg
 
-    def visit_SetAt(self, stmt: gtir.SetAt, sdfg: dace.SDFG, state: dace.SDFGState) -> None:
+    def visit_SetAt(
+        self, stmt: gtir.SetAt, sdfg: dace.SDFG, state: dace.SDFGState
+    ) -> dace.SDFGState:
         """Visits a `SetAt` statement expression and writes the local result to some external storage.
 
         Each statement expression results in some sort of dataflow gragh writing to temporary storage.
         The translation of `SetAt` ensures that the result is written back to the target external storage.
+
+        Returns:
+          The SDFG head state, eventually updated if the target write requires a new state.
         """
 
-        expr_nodes = self._visit_expression(stmt.expr, sdfg, state)
+        temp_fields = self._visit_expression(stmt.expr, sdfg, state)
 
         # the target expression could be a `SymRef` to an output node or a `make_tuple` expression
         # in case the statement returns more than one field
-        target_nodes = self._visit_expression(stmt.target, sdfg, state)
+        target_fields = self._visit_expression(stmt.target, sdfg, state, use_temp=False)
 
         # convert domain expression to dictionary to ease access to dimension boundaries
         domain = dace_fieldview_util.get_domain_ranges(stmt.domain)
 
-        for expr_node, target_node in zip(expr_nodes, target_nodes, strict=True):
-            target_array = sdfg.arrays[target_node.data]
-            assert not target_array.transient
-            target_symbol_type = self.global_symbols[target_node.data]
+        expr_input_args = {
+            str(sym.id)
+            for sym in eve.walk_values(stmt.expr).if_isinstance(gtir.SymRef)
+            if str(sym.id) in sdfg.arrays
+        }
```

Comment:
or
```suggestion
        expr_input_args = eve.walk_values(stmt.expr).if_isinstance(gtir.SymRef).filter(lambda sym: str(sym.id) in sdfg.arrays).map(lambda sym: str(sym.id)).to_set()
```
or just leave at is...